In [1]:
import pandas as pd
import regex as re
from datetime import datetime
import numpy as np

## Method1
Converted Excel data file into CSV and then importing into Python 

In [2]:
Exit_Notes=pd.read_csv('../data/Exit_Notes.csv',encoding='cp1252')

In [3]:
Exit_Notes.head()


POE  Child ID        DOB Notification Date Initial Eligibility Date  \
0  UC    471738  2/28/2019        10/17/2019               10/29/2019   
1  FT    449030  7/22/2018         11/7/2018                12/3/2018   
2  FT    463311   7/9/2017         5/24/2019                6/19/2019   
3  FT    463313   7/9/2017         5/24/2019                6/19/2019   
4  FT    469194  8/31/2017          9/4/2019                10/2/2019   

  Initial IFSP Date Latest IFSP Date                              Exit Reason  \
0        10/31/2019        5/12/2020                    618 - Parent withdraw   
1        12/18/2018         3/4/2021                 618 - Moved out of state   
2          7/2/2019        6/22/2020  618 - Part B eligibility not determined   
3          7/2/2019        6/22/2020  618 - Part B eligibility not determined   
4        10/11/2019        5/26/2020  618 - Part B eligibility not determined   

   Exit Date  Days from I-IFSP to Exit  \
0  9/21/2020                       326   
1   3/6/2021                       809   
2   7/8/2020                       372   
3   7/8/2020                       372   
4  8/30/2020                       324   

                                           Exit Note            Exit category  \
0            * No Exit ECO available due to COVID-19  NO exit dates or reopen   
1  ***family moving out of state March 2021, and ...  NO exit dates or reopen   
2  **Completed the transition planning conference...                      NaN   
3  **Completed the transition planning conference...                      NaN   
4  **Due to the impact of the COVID-19 pandemic a...                      NaN   

   Unnamed: 12  Unnamed: 13  
0          NaN          NaN  
1          NaN       1395.5  
2          NaN          NaN  
3          NaN          NaN  
4          NaN          NaN

In [4]:
Exit_Notes[['POE','Child ID','DOB','Initial IFSP Date','Exit Reason','Exit Date',
            'Days from I-IFSP to Exit','Exit Note']]

POE  Child ID         DOB Initial IFSP Date  \
0     UC    471738   2/28/2019        10/31/2019   
1     FT    449030   7/22/2018        12/18/2018   
2     FT    463311    7/9/2017          7/2/2019   
3     FT    463313    7/9/2017          7/2/2019   
4     FT    469194   8/31/2017        10/11/2019   
...   ..       ...         ...               ...   
5577  MD    491609  12/22/2018         11/9/2020   
5578  GN    491743    6/5/2018         11/9/2020   
5579  ET    491879   4/24/2019         11/4/2020   
5580  SC    492267    6/1/2018        11/24/2020   
5581  SW    492408    8/6/2020        11/20/2020   

                                  Exit Reason  Exit Date  \
0                       618 - Parent withdraw  9/21/2020   
1                    618 - Moved out of state   3/6/2021   
2     618 - Part B eligibility not determined   7/8/2020   
3     618 - Part B eligibility not determined   7/8/2020   
4     618 - Part B eligibility not determined  8/30/2020   
...                                       ...        ...   
5577   618 - Attempts to contact unsuccessful  5/17/2021   
5578  618 - Part B eligibility not determined   6/4/2021   
5579                    618 - Parent withdraw   5/6/2021   
5580  618 - Part B eligibility not determined  5/31/2021   
5581   618 - Attempts to contact unsuccessful  5/20/2021   

      Days from I-IFSP to Exit  \
0                          326   
1                          809   
2                          372   
3                          372   
4                          324   
...                        ...   
5577                       189   
5578                       207   
5579                       183   
5580                       188   
5581                       181   

                                              Exit Note  
0               * No Exit ECO available due to COVID-19  
1     ***family moving out of state March 2021, and ...  
2     **Completed the transition planning conference...  
3     **Completed the transition planning conference...  
4     **Due to the impact of the COVID-19 pandemic a...  
...                                                 ...  
5577                                                NaN  
5578                                                NaN  
5579                                                NaN  
5580                                                NaN  
5581                                                NaN  

[5582 rows x 8 columns]

## Method 2
Reading the Excel file directly into Python

In [5]:
Data_File = pd.read_excel('../data/TEIS Exit Data for NSS.xlsx', sheet_name = None) #imports all data

In [6]:
 Data_File.keys()

dict_keys(['Reported Data', 'Reported Totals', 'ECO with Exit20-21', 'Exit Notes', 'Terms-Acronyms'])

In [7]:
Data_File.values()

dict_values([                     Child Outcome Summay FY 2020-21 Unnamed: 1 Unnamed: 2  \
0   POE comparison to State Target on Child Outcomes        NaN        NaN   
1                                                NaN        NaN        NaN   
2                                          Outcome A        NaN        NaN   
3                                                NaN    OC1 - a    OC1 - b   
4                                                 ET          0        155   
5                                                 FT          0        121   
6                                                 GN          8        294   
7                                                 MD          4        315   
8                                                 NW          0         68   
9                                                 SC          1        207   
10                                                SE          0         81   
11                                                S

## Reading a Sheet from Excel 

In [8]:
Exit_Notes=Data_File['Exit Notes']
Exit_Notes_Original=Data_File['Exit Notes']  # Copy of original Data Sheet 

In [98]:
ECO_with_Exit20_21=Data_File['ECO with Exit20-21']
ECO_with_Exit20_21_Original=Data_File['ECO with Exit20-21'] # Copy of original Data Sheet

In [9]:
Exit_Notes

POE  Child ID        DOB Notification Date Initial Eligibility Date  \
0     UC    471738 2019-02-28        2019-10-17               2019-10-29   
1     FT    449030 2018-07-22        2018-11-07               2018-12-03   
2     FT    463311 2017-07-09        2019-05-24               2019-06-19   
3     FT    463313 2017-07-09        2019-05-24               2019-06-19   
4     FT    469194 2017-08-31        2019-09-04               2019-10-02   
...   ..       ...        ...               ...                      ...   
5577  MD    491609 2018-12-22        2020-10-20               2020-10-30   
5578  GN    491743 2018-06-05        2020-10-22               2020-11-05   
5579  ET    491879 2019-04-24        2020-10-26               2020-11-02   
5580  SC    492267 2018-06-01        2020-10-30               2020-11-17   
5581  SW    492408 2020-08-06        2020-11-03               2020-11-17   

     Initial IFSP Date Latest IFSP Date  \
0           2019-10-31       2020-05-12   
1           2018-12-18       2021-03-04   
2           2019-07-02       2020-06-22   
3           2019-07-02       2020-06-22   
4           2019-10-11       2020-05-26   
...                ...              ...   
5577        2020-11-09       2021-03-05   
5578        2020-11-09       2021-05-04   
5579        2020-11-04       2021-04-15   
5580        2020-11-24       2021-04-15   
5581        2020-11-20       2020-11-20   

                                  Exit Reason  Exit Date  \
0                       618 - Parent withdraw 2020-09-21   
1                    618 - Moved out of state 2021-03-06   
2     618 - Part B eligibility not determined 2020-07-08   
3     618 - Part B eligibility not determined 2020-07-08   
4     618 - Part B eligibility not determined 2020-08-30   
...                                       ...        ...   
5577   618 - Attempts to contact unsuccessful 2021-05-17   
5578  618 - Part B eligibility not determined 2021-06-04   
5579                    618 - Parent withdraw 2021-05-06   
5580  618 - Part B eligibility not determined 2021-05-31   
5581   618 - Attempts to contact unsuccessful 2021-05-20   

      Days from I-IFSP to Exit  \
0                          326   
1                          809   
2                          372   
3                          372   
4                          324   
...                        ...   
5577                       189   
5578                       207   
5579                       183   
5580                       188   
5581                       181   

                                              Exit Note  
0               * No Exit ECO available due to COVID-19  
1     ***family moving out of state March 2021, and ...  
2     **Completed the transition planning conference...  
3     **Completed the transition planning conference...  
4     **Due to the impact of the COVID-19 pandemic a...  
...                                                 ...  
5577                                                NaN  
5578                                                NaN  
5579                                                NaN  
5580                                                NaN  
5581                                                NaN  

[5582 rows x 11 columns]

In [10]:
Exit_Notes=Exit_Notes[['POE','Child ID','DOB','Initial IFSP Date','Exit Reason','Exit Date',
            'Days from I-IFSP to Exit','Exit Note']]

## Renaming Columns

In [11]:
Exit_Notes=Exit_Notes.rename(columns={'Initial IFSP Date':'Init_IFSP_Dt','Child ID':'ID','Exit Reason':'Exit_Rsn',
                          'Exit Date':'Exit_Dt','Days from I-IFSP to Exit':'Service_Days'})

In [12]:
Exit_Notes.loc[Exit_Notes['Exit Note'].isna()]

POE      ID        DOB Init_IFSP_Dt  \
2251  SE  421699 2017-07-13   2017-09-01   
2252  ET  421707 2017-07-10   2017-09-05   
2253  UC  422055 2017-07-24   2017-08-30   
2254  ET  422065 2017-07-12   2017-09-18   
2255  UC  422838 2017-07-03   2017-09-19   
...   ..     ...        ...          ...   
5577  MD  491609 2018-12-22   2020-11-09   
5578  GN  491743 2018-06-05   2020-11-09   
5579  ET  491879 2019-04-24   2020-11-04   
5580  SC  492267 2018-06-01   2020-11-24   
5581  SW  492408 2020-08-06   2020-11-20   

                                     Exit_Rsn    Exit_Dt  Service_Days  \
2251  618 - Part B eligibility not determined 2020-07-12          1045   
2252  618 - Part B eligibility not determined 2020-07-09          1038   
2253  618 - Part B eligibility not determined 2020-07-23          1058   
2254  618 - Part B eligibility not determined 2020-07-11          1027   
2255  618 - Part B eligibility not determined 2020-07-02          1017   
...                                       ...        ...           ...   
5577   618 - Attempts to contact unsuccessful 2021-05-17           189   
5578  618 - Part B eligibility not determined 2021-06-04           207   
5579                    618 - Parent withdraw 2021-05-06           183   
5580  618 - Part B eligibility not determined 2021-05-31           188   
5581   618 - Attempts to contact unsuccessful 2021-05-20           181   

     Exit Note  
2251       NaN  
2252       NaN  
2253       NaN  
2254       NaN  
2255       NaN  
...        ...  
5577       NaN  
5578       NaN  
5579       NaN  
5580       NaN  
5581       NaN  

[3331 rows x 8 columns]

## Create a new column Actual Service Days

In [13]:
 Exit_Notes.insert(7,"Actual_Service_Days",0, False)

In [14]:
Exit_Notes.head()

POE      ID        DOB Init_IFSP_Dt  \
0  UC  471738 2019-02-28   2019-10-31   
1  FT  449030 2018-07-22   2018-12-18   
2  FT  463311 2017-07-09   2019-07-02   
3  FT  463313 2017-07-09   2019-07-02   
4  FT  469194 2017-08-31   2019-10-11   

                                  Exit_Rsn    Exit_Dt  Service_Days  \
0                    618 - Parent withdraw 2020-09-21           326   
1                 618 - Moved out of state 2021-03-06           809   
2  618 - Part B eligibility not determined 2020-07-08           372   
3  618 - Part B eligibility not determined 2020-07-08           372   
4  618 - Part B eligibility not determined 2020-08-30           324   

   Actual_Service_Days                                          Exit Note  
0                    0            * No Exit ECO available due to COVID-19  
1                    0  ***family moving out of state March 2021, and ...  
2                    0  **Completed the transition planning conference...  
3                    0  **Completed the transition planning conference...  
4                    0  **Due to the impact of the COVID-19 pandemic a...

## Taking a copy of original Df

In [15]:
Exit_Notes_Copy = Exit_Notes.copy()

In [16]:
Exit_Notes_Copy.head()

POE      ID        DOB Init_IFSP_Dt  \
0  UC  471738 2019-02-28   2019-10-31   
1  FT  449030 2018-07-22   2018-12-18   
2  FT  463311 2017-07-09   2019-07-02   
3  FT  463313 2017-07-09   2019-07-02   
4  FT  469194 2017-08-31   2019-10-11   

                                  Exit_Rsn    Exit_Dt  Service_Days  \
0                    618 - Parent withdraw 2020-09-21           326   
1                 618 - Moved out of state 2021-03-06           809   
2  618 - Part B eligibility not determined 2020-07-08           372   
3  618 - Part B eligibility not determined 2020-07-08           372   
4  618 - Part B eligibility not determined 2020-08-30           324   

   Actual_Service_Days                                          Exit Note  
0                    0            * No Exit ECO available due to COVID-19  
1                    0  ***family moving out of state March 2021, and ...  
2                    0  **Completed the transition planning conference...  
3                    0  **Completed the transition planning conference...  
4                    0  **Due to the impact of the COVID-19 pandemic a...

## Handling blank Exit Notes

In [17]:
for index,row  in Exit_Notes.iterrows():
   if pd.isna(row['Exit Note']) :
        #Exit_Notes.Actual_Service_Days=Exit_Notes.Service_Days
        Exit_Notes.loc[index,'Actual_Service_Days']= Exit_Notes.loc[index,'Service_Days']
               #print(row['Exit Note'],row['Service_Days'],row['Actual_Service_Days'])
    


In [18]:
Exit_Notes_Copy.groupby('Actual_Service_Days')['ID'].count()

Actual_Service_Days
0    5582
Name: ID, dtype: int64

In [19]:
(Exit_Notes['Actual_Service_Days'] != 0).sum()

3331

In [20]:
Exit_Notes=Exit_Notes.rename(columns={'Exit Note':'Exit_Note'})
Exit_Notes_Copy=Exit_Notes_Copy.rename(columns={'Exit Note':'Exit_Note'})

In [21]:
Exit_Notes1=Exit_Notes.loc[Exit_Notes['Actual_Service_Days'] == 0]
Exit_Notes1.head()

POE      ID        DOB Init_IFSP_Dt  \
0  UC  471738 2019-02-28   2019-10-31   
1  FT  449030 2018-07-22   2018-12-18   
2  FT  463311 2017-07-09   2019-07-02   
3  FT  463313 2017-07-09   2019-07-02   
4  FT  469194 2017-08-31   2019-10-11   

                                  Exit_Rsn    Exit_Dt  Service_Days  \
0                    618 - Parent withdraw 2020-09-21           326   
1                 618 - Moved out of state 2021-03-06           809   
2  618 - Part B eligibility not determined 2020-07-08           372   
3  618 - Part B eligibility not determined 2020-07-08           372   
4  618 - Part B eligibility not determined 2020-08-30           324   

   Actual_Service_Days                                          Exit_Note  
0                    0            * No Exit ECO available due to COVID-19  
1                    0  ***family moving out of state March 2021, and ...  
2                    0  **Completed the transition planning conference...  
3                    0  **Completed the transition planning conference...  
4                    0  **Due to the impact of the COVID-19 pandemic a...

## Formatting Jupyter Notebook Cell

In [22]:
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:100% !important; }</style>"))

In [23]:
 display(HTML("<style>.output_result { max-width:100% !important; }</style>"))

In [24]:
pd.options.display.max_rows
pd.set_option('display.max_colwidth', None)
pd.reset_option('max_columns')

## Cleaning Begins 

In [25]:

Exit_Notes1[Exit_Notes1['Exit_Note'].str.contains('([0-9]+[\/\-][0-9]+[\/\-][0-9]{2}\D)')==True]


C:\Users\Anand\AppData\Local\Temp/ipykernel_21912/1139523904.py:1: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  Exit_Notes1[Exit_Notes1['Exit_Note'].str.contains('([0-9]+[\/\-][0-9]+[\/\-][0-9]{2}\D)')==True]


POE      ID        DOB Init_IFSP_Dt  \
6     NW  460887 2017-07-19   2019-05-29   
15    GN  474532 2019-11-26   2020-01-15   
19    GN  485954 2020-01-27   2020-08-07   
20    ET  473290 2018-10-19   2019-12-09   
21    GN  476958 2018-10-17   2020-08-24   
...   ..     ...        ...          ...   
2219  SC  460481 2019-01-07   2019-05-17   
2233  SE  428532 2017-07-13   2018-01-26   
2235  FT  471858 2017-09-19   2019-11-21   
2243  ET  463610 2017-08-09   2019-08-12   
2248  GN  450369 2017-08-16   2018-12-19   

                                                    Exit_Rsn    Exit_Dt  \
6     618 - Not eligible for Part B – Exit with no referrals 2020-07-18   
15                                     618 - Parent withdraw 2021-04-20   
19                                     618 - Parent withdraw 2021-05-24   
20                                     618 - Parent withdraw 2021-05-24   
21                                     618 - Parent withdraw 2021-05-25   
...                                                      ...        ...   
2219                                   618 - Parent withdraw 2020-07-07   
2233                 618 - Part B eligibility not determined 2020-07-12   
2235                                   618 - Part B eligible 2020-09-18   
2243                  618 - Attempts to contact unsuccessful 2020-08-08   
2248                 618 - Part B eligibility not determined 2020-08-15   

      Service_Days  Actual_Service_Days  \
6              416                    0   
15             461                    0   
19             290                    0   
20             532                    0   
21             274                    0   
...            ...                  ...   
2219           417                    0   
2233           898                    0   
2235           302                    0   
2243           362                    0   
2248           605                    0   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             Exit_Note  
6     *Continuation of Transition Meeting Conference note of 11-5-19 -- Non LEA service options in the community such as private preschool and daycare were discussed. Speech therapy paid by TennCare can continue after age three. Transition brochure was given. LEA options for a three year old were explained. LEA Process for determining eligibility was explained. Testing will be completed and if eligible Individualized Education Plan (IEP) will be written prior to child's third birthday. Plans are for IP to be written prior to the end of this school year. SC will visit child at Early Head Start this morning for monthly Targeted Case Management (TCM) visit.  
15                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           04/20/21- parent withdraw  
19                                                             

## Cleaning  Unsuccessful 
    for index,row  in Exit_Notes1.iterrows():
      
      if re.search("[0-9]+[\/\-][0-9]+[\/\-]([0-9]{2})\D", row['Exit_Note']) :
         row['Exit_Note']=re.sub(r'[0-9]+[\/\-][0-9]+[\/\-]', r'[0-9]+[\/\-][0-9]+[\/\-]20', row['Exit_Note'])
         print("***",row['Exit_Note']) 
    #Exit_Notes['Exit_Note'].replace(to_replace=r"([0-9]+[\/\-][0-9]+[\/\-][0-9]{2}\D)",value=r'[0-9]+[\/\\-][0-9]+[\/\-]20[0-9]{2}\D',regex=True).head(100)



### Calculating Actual Days of Service

In [26]:
Exit_Reopen=Exit_Notes1[Exit_Notes1['Exit_Note'].str.contains('Exit Date:|Reopened:|Exit Reason:')==True]

In [27]:
Exit_Reopen.head()

POE      ID        DOB Init_IFSP_Dt  \
5   UC  473474 2017-11-05   2020-05-06   
8   GN  469243 2019-07-26   2019-10-02   
12  SE  436339 2018-04-05   2019-02-13   
16  ET  474102 2018-05-11   2020-04-09   
21  GN  476958 2018-10-17   2020-08-24   

                                   Exit_Rsn    Exit_Dt  Service_Days  \
5   618 - Part B eligibility not determined 2020-11-04           182   
8                     618 - Parent withdraw 2021-02-11           498   
12  618 - Part B eligibility not determined 2021-04-01           778   
16  618 - Part B eligibility not determined 2021-05-10           396   
21                    618 - Parent withdraw 2021-05-25           274   

    Actual_Service_Days  \
5                     0   
8                     0   
12                    0   
16                    0   
21                    0   

                                                                                                                                                                                                                                                         Exit_Note  
5                                         **LEA county on the Transition page and demographics page do not match due to child moving out of county following the transition meeting. Reopened: 04/06/2020 Exit Date: 11/26/2019 Exit Reason: Ineligible for Part C  
8                                           02-11-2021 – PWN signed and returned for voluntary closure/declined services. Family made aware they can reopen the file if they choose. Reopened: 02/05/2021 Exit Date: 04/07/2020 Exit Reason: 618 - Parent withdraw  
12  04/01/2021- File closed business day prior to child's birthday. Parents declined TPC with LEA and plan to continue with private preschool and feeding therapy through insurance. Reopened: 01/11/2019 Exit Date: 06/17/2018 Exit Reason: Ineligible for Part C  
16                                                                                                05/10/2021 TPC held then parents withdrew from testing and thus part B not determined. Reopened: 02/28/2020 Exit Date: 12/05/2019 Exit Reason: Unable to contact  
21                                                                                                                                                             05/25/21- parent withdraw Reopened: 08/04/2020 Exit Date: 03/09/2020 Exit Reason: Unable to contact

In [28]:
Exit=Exit_Reopen['Exit_Note'].str.extractall(pat ='Exit Date: (\d{2}\/\d{2}\/\d{4})').unstack()

In [29]:
Reopen=Exit_Reopen['Exit_Note'].str.extractall(pat ='Reopened: (\d{2}\/\d{2}\/\d{4})').unstack()

In [30]:
Exit.head()

0                         
match           0    1    2    3    4    5
5      11/26/2019  NaN  NaN  NaN  NaN  NaN
8      04/07/2020  NaN  NaN  NaN  NaN  NaN
12     06/17/2018  NaN  NaN  NaN  NaN  NaN
16     12/05/2019  NaN  NaN  NaN  NaN  NaN
21     03/09/2020  NaN  NaN  NaN  NaN  NaN

In [31]:
from functools import reduce

Exit_Reopen_Unstack=reduce(lambda x,y: pd.merge(x,y, left_index=True, right_index=True, how='inner'), [Exit_Reopen, Exit, Reopen])

C:\Users\Anand\AppData\Local\Temp/ipykernel_21912/3418294049.py:3: FutureWarning: merging between different levels is deprecated and will be removed in a future version. (1 levels on the left,2 on the right)
  Exit_Reopen_Unstack=reduce(lambda x,y: pd.merge(x,y, left_index=True, right_index=True, how='inner'), [Exit_Reopen, Exit, Reopen])


In [32]:
Exit_Reopen_Unstack.head(1)

POE      ID        DOB Init_IFSP_Dt  \
5  UC  473474 2017-11-05   2020-05-06   

                                  Exit_Rsn    Exit_Dt  Service_Days  \
5  618 - Part B eligibility not determined 2020-11-04           182   

   Actual_Service_Days  \
5                    0   

                                                                                                                                                                                                                  Exit_Note  \
5  **LEA county on the Transition page and demographics page do not match due to child moving out of county following the transition meeting. Reopened: 04/06/2020 Exit Date: 11/26/2019 Exit Reason: Ineligible for Part C   

     (0, 0)_x  ... (0, 2)_x (0, 3)_x (0, 4)_x (0, 5)_x      (0, 0) (0, 1)  \
5  11/26/2019  ...      NaN      NaN      NaN      NaN  04/06/2020    NaN   

  (0, 2) (0, 3) (0, 4) (0, 5)  
5    NaN    NaN    NaN    NaN  

[1 rows x 21 columns]

In [33]:
Exit_Reopen_Unstack_subset=Exit_Reopen_Unstack.drop(columns=['POE','Exit_Rsn','Exit_Note'])

In [34]:
Exit_Reopen_Unstack_subset.head(500)

ID        DOB Init_IFSP_Dt    Exit_Dt  Service_Days  \
5    473474 2017-11-05   2020-05-06 2020-11-04           182   
8    469243 2019-07-26   2019-10-02 2021-02-11           498   
12   436339 2018-04-05   2019-02-13 2021-04-01           778   
16   474102 2018-05-11   2020-04-09 2021-05-10           396   
21   476958 2018-10-17   2020-08-24 2021-05-25           274   
..      ...        ...          ...        ...           ...   
969  436968 2017-07-25   2019-08-02 2020-07-24           357   
970  433438 2017-10-31   2018-05-10 2020-10-26           900   
971  478402 2018-06-17   2020-04-30 2020-11-30           214   
972  479629 2020-01-27   2020-05-21 2020-11-30           193   
973  485520 2018-12-31   2020-08-17 2021-04-09           235   

     Actual_Service_Days    (0, 0)_x    (0, 1)_x    (0, 2)_x (0, 3)_x  \
5                      0  11/26/2019         NaN         NaN      NaN   
8                      0  04/07/2020         NaN         NaN      NaN   
12                     0  06/17/2018         NaN         NaN      NaN   
16                     0  12/05/2019         NaN         NaN      NaN   
21                     0  03/09/2020         NaN         NaN      NaN   
..                   ...         ...         ...         ...      ...   
969                    0  02/17/2020  06/04/2018         NaN      NaN   
970                    0  09/12/2018         NaN         NaN      NaN   
971                    0  02/26/2020         NaN         NaN      NaN   
972                    0  03/16/2020         NaN         NaN      NaN   
973                    0  01/29/2021  01/29/2021  08/06/2020      NaN   

    (0, 4)_x (0, 5)_x      (0, 0)      (0, 1)      (0, 2) (0, 3) (0, 4) (0, 5)  
5        NaN      NaN  04/06/2020         NaN         NaN    NaN    NaN    NaN  
8        NaN      NaN  02/05/2021         NaN         NaN    NaN    NaN    NaN  
12       NaN      NaN  01/11/2019         NaN         NaN    NaN    NaN    NaN  
16       NaN      NaN  02/28/2020         NaN         NaN    NaN    NaN    NaN  
21       NaN      NaN  08/04/2020         NaN         NaN    NaN    NaN    NaN  
..       ...      ...         ...         ...         ...    ...    ...    ...  
969      NaN      NaN  03/16/2020  06/26/2019         NaN    NaN    NaN    NaN  
970      NaN      NaN  03/16/2020         NaN         NaN    NaN    NaN    NaN  
971      NaN      NaN  03/17/2020         NaN         NaN    NaN    NaN    NaN  
972      NaN      NaN  03/17/2020         NaN         NaN    NaN    NaN    NaN  
973      NaN      NaN  03/17/2021  02/02/2021  08/14/2020    NaN    NaN    NaN  

[500 rows x 18 columns]

In [35]:
mapping={'(0, 0)_x':'ExDT1','(0, 1)_x':'ExDT2','(0, 2)_x':'ExDT3','(0, 3)_x':'ExDT4','(0, 4)_x':'ExDT5','(0, 5)_x':'ExDT6'}
         
Exit_Reopen_Unstack_subset = Exit_Reopen_Unstack_subset.rename(columns=mapping)

In [36]:
mapping={(0, 0):'ReDT1',(0, 1):'ReDT2',(0, 2):'ReDT3',(0, 3):'ReDT4',(0, 4):'ReDT5',(0, 5):'ReDT6'}
Exit_Reopen_Unstack_subset = Exit_Reopen_Unstack_subset.rename(columns=mapping)

In [37]:
Exit_Reopen_Unstack_subset.head()


ID        DOB Init_IFSP_Dt    Exit_Dt  Service_Days  \
5   473474 2017-11-05   2020-05-06 2020-11-04           182   
8   469243 2019-07-26   2019-10-02 2021-02-11           498   
12  436339 2018-04-05   2019-02-13 2021-04-01           778   
16  474102 2018-05-11   2020-04-09 2021-05-10           396   
21  476958 2018-10-17   2020-08-24 2021-05-25           274   

    Actual_Service_Days       ExDT1 ExDT2 ExDT3 ExDT4 ExDT5 ExDT6       ReDT1  \
5                     0  11/26/2019   NaN   NaN   NaN   NaN   NaN  04/06/2020   
8                     0  04/07/2020   NaN   NaN   NaN   NaN   NaN  02/05/2021   
12                    0  06/17/2018   NaN   NaN   NaN   NaN   NaN  01/11/2019   
16                    0  12/05/2019   NaN   NaN   NaN   NaN   NaN  02/28/2020   
21                    0  03/09/2020   NaN   NaN   NaN   NaN   NaN  08/04/2020   

   ReDT2 ReDT3 ReDT4 ReDT5 ReDT6  
5    NaN   NaN   NaN   NaN   NaN  
8    NaN   NaN   NaN   NaN   NaN  
12   NaN   NaN   NaN   NaN   NaN  
16   NaN   NaN   NaN   NaN   NaN  
21   NaN   NaN   NaN   NaN   NaN

## For Analysis
Exit_Reopen_Unstack_subset.to_csv("../data/Exit_Reopen_Unstack_subset.csv", index = False)

In [38]:
for index,row  in Exit_Reopen_Unstack_subset.iterrows():
    #if (row['ID'] == 461388):
        IFSP=row['Init_IFSP_Dt'] 
        EXITDT=row['Exit_Dt']
        SD=row['Service_Days']
        ASD=row['Actual_Service_Days']
        print("#",SD)
        print("#",ASD)
        Exit= [row['ExDT6'],row['ExDT5'],row['ExDT4'],row['ExDT3'],row['ExDT2'],row['ExDT1']]
        Reopen= [row['ReDT6'],row['ReDT5'],row['ReDT4'],row['ReDT3'],row['ReDT2'],row['ReDT1']]
     
        E=[]
        for item in Exit:
            item=pd.to_datetime(item, format='%m/%d/%Y')
            if (pd.notnull(item)) & (item >= IFSP ) :  # no early dates thanIFSP
                if (len(E)==0) & (item < EXITDT) :
                    E.append(item)
                    print("*",SD)
                    print("*",ASD)
                elif (len(E) >= 1) & (item < EXITDT):  # No Duplicates between exit date in exit note and Final Exit date
                    if item > E[-1]: # NO duplicate entries between exit dates in exit notes
                        E.append(item)
                        print("***",SD)
                        print("***",ASD)
                    else:
                        E=[]
                        print("**",SD)
                        print("**",ASD)
                        break
                else :
                    continue
        print(len(E))
        R=[]
        for item in Reopen:
            item=pd.to_datetime(item, format='%m/%d/%Y')
            if (pd.notnull(item)) & (item >= IFSP ):
                if len(R)==0:
                    R.append(item)
                elif (len(R) >= 1) & (item > R[-1]):
                     R.append(item)
                else:
                    R=[]
                    break
        print(len(R))
        if  len(E) == 0 & len(R) == 0 :
            Exit_Reopen_Unstack_subset.loc[index,'Actual_Service_Days']=(EXITDT-IFSP).days 
        elif len(E) == 1 & len(R) == 1 :
             Exit_Reopen_Unstack_subset.loc[index,'Actual_Service_Days']=(E[0]-IFSP).days + (EXITDT-R[0]).days
        elif len(E) == 2 & len(R) == 2 : 
             Exit_Reopen_Unstack_subset.loc[index,'Actual_Service_Days']=(E[0]-IFSP).days+(E[1]-R[0]).days+(EXITDT-R[1]).days
        elif len(E) == 3 & len(R) == 3 : 
             Exit_Reopen_Unstack_subset.loc[index,'Actual_Service_Days']=(E[0]-IFSP).days+(E[1]-R[0]).days+(E[2]-R[1]).days+(EXITDT-R[2]).days
        elif len(E) == 4 & len(R) == 4:
             Exit_Reopen_Unstack_subset.loc[index,'Actual_Service_Days']=(E[0]-IFSP).days +(E[1]-R[0]).days+(E[2]-R[1]).days+(E[3]-R[2]).days+(EXITDT-R[3]).days
        elif  len(E) == 5 & len(R) == 5:
             Exit_Reopen_Unstack_subset.loc[index,'Actual_Service_Days']=(E[0]-IFSP).days+(E[1]-R[0]).days+(E[2]-R[1]).days+(E[3]-R[2]).days+(E[4]-R[3]).days+(EXITDT-R[4]).days
        elif len(E) == 6 & len(R) == 6: 
            Exit_Reopen_Unstack_subset.loc[index,'Actual_Service_Days']=(E[0]-IFSP).days+(E[1]-R[0]).days+(E[2]-R[1]).days+(E[3]-R[2]).days+(E[4]-R[3]).days+(E[5]-R[4]).days+(EXITDT-R[5]).days
        else:
            Exit_Reopen_Unstack_subset.loc[index,'Actual_Service_Days']=(EXITDT-IFSP).days

# 182
# 0
0
0
# 498
# 0
* 498
* 0
1
1
# 778
# 0
0
0
# 396
# 0
0
0
# 274
# 0
0
0
# 902
# 0
* 902
* 0
** 902
** 0
0
4
# 185
# 0
0
0
# 314
# 0
* 314
* 0
1
1
# 281
# 0
0
0
# 584
# 0
* 584
* 0
** 584
** 0
0
1
# 438
# 0
0
0
# 517
# 0
0
0
# 399
# 0
* 399
* 0
1
1
# 518
# 0
* 518
* 0
1
1
# 856
# 0
0
0
# 426
# 0
0
0
# 997
# 0
* 997
* 0
*** 997
*** 0
** 997
** 0
0
3
# 997
# 0
* 997
* 0
*** 997
*** 0
2
2
# 620
# 0
0
0
# 215
# 0
* 215
* 0
1
1
# 929
# 0
* 929
* 0
*** 929
*** 0
2
2
# 184
# 0
0
0
# 932
# 0
0
0
# 661
# 0
0
0
# 671
# 0
* 671
* 0
*** 671
*** 0
2
2
# 785
# 0
0
0
# 632
# 0
* 632
* 0
*** 632
*** 0
*** 632
*** 0
*** 632
*** 0
4
4
# 306
# 0
0
0
# 243
# 0
0
0
# 881
# 0
0
0
# 726
# 0
* 726
* 0
1
1
# 460
# 0
* 460
* 0
1
1
# 612
# 0
0
0
# 715
# 0
0
0
# 315
# 0
0
0
# 328
# 0
0
0
# 672
# 0
0
0
# 1001
# 0
* 1001
* 0
1
1
# 441
# 0
* 441
* 0
1
1
# 196
# 0
* 196
* 0
1
1
# 539
# 0
0
0
# 601
# 0
* 601
* 0
1
1
# 212
# 0
0
0
# 767
# 0
* 767
* 0
*** 767
*** 0
2
2
# 742
# 0
* 742
* 0
1
1
# 742
# 0
* 742
* 0


0
# 638
# 0
* 638
* 0
*** 638
*** 0
2
1
# 541
# 0
0
0
# 1025
# 0
* 1025
* 0
1
1
# 607
# 0
0
0
# 316
# 0
0
0
# 511
# 0
* 511
* 0
1
1
# 246
# 0
0
0
# 758
# 0
0
0
# 609
# 0
0
0
# 305
# 0
0
0
# 316
# 0
0
0
# 286
# 0
0
0
# 525
# 0
0
0
# 934
# 0
0
0
# 962
# 0
* 962
* 0
** 962
** 0
0
2
# 570
# 0
0
0
# 960
# 0
0
0
# 851
# 0
0
0
# 242
# 0
0
0
# 540
# 0
* 540
* 0
1
1
# 320
# 0
0
0
# 276
# 0
* 276
* 0
1
1
# 497
# 0
0
1
# 295
# 0
0
0
# 364
# 0
0
0
# 649
# 0
* 649
* 0
1
1
# 468
# 0
0
1
# 564
# 0
* 564
* 0
1
1
# 871
# 0
0
0
# 298
# 0
* 298
* 0
1
1
# 418
# 0
* 418
* 0
1
1
# 241
# 0
0
0
# 771
# 0
* 771
* 0
*** 771
*** 0
*** 771
*** 0
3
3
# 541
# 0
0
0
# 717
# 0
0
0
# 873
# 0
* 873
* 0
1
1
# 182
# 0
0
0
# 226
# 0
0
0
# 188
# 0
0
0
# 930
# 0
0
0
# 931
# 0
0
0
# 602
# 0
* 602
* 0
1
1
# 791
# 0
* 791
* 0
1
1
# 328
# 0
* 328
* 0
1
1
# 588
# 0
* 588
* 0
1
1
# 357
# 0
* 357
* 0
1
1
# 900
# 0
* 900
* 0
1
1
# 214
# 0
0
0
# 193
# 0
0
0
# 235
# 0
* 235
* 0
** 235
** 0
0
2
# 714
# 0
* 714
* 0
** 714
** 0
0
2
# 31

# 341
# 0
0
0
# 746
# 0
* 746
* 0
1
1
# 472
# 0
* 472
* 0
1
1
# 198
# 0
0
0
# 465
# 0
0
1
# 875
# 0
0
0
# 738
# 0
0
0
# 781
# 0
0
0
# 280
# 0
0
0
# 196
# 0
0
0
# 359
# 0
* 359
* 0
1
1
# 838
# 0
* 838
* 0
** 838
** 0
0
2
# 261
# 0
0
0
# 1029
# 0
* 1029
* 0
** 1029
** 0
0
2
# 1012
# 0
* 1012
* 0
*** 1012
*** 0
2
2
# 186
# 0
0
0
# 512
# 0
0
0
# 371
# 0
0
0
# 360
# 0
0
0
# 530
# 0
0
0
# 684
# 0
* 684
* 0
** 684
** 0
0
3
# 227
# 0
0
0
# 796
# 0
0
0
# 724
# 0
0
0
# 537
# 0
0
0
# 198
# 0
0
0
# 206
# 0
0
1
# 255
# 0
0
0
# 530
# 0
0
1
# 362
# 0
* 362
* 0
1
1
# 868
# 0
* 868
* 0
1
1
# 436
# 0
0
0
# 207
# 0
0
0
# 1036
# 0
* 1036
* 0
1
1
# 760
# 0
0
0
# 843
# 0
0
0
# 481
# 0
0
0
# 540
# 0
* 540
* 0
1
1
# 711
# 0
0
0
# 360
# 0
0
0
# 479
# 0
* 479
* 0
1
1
# 590
# 0
0
0
# 837
# 0
0
0
# 938
# 0
0
0
# 764
# 0
0
0
# 625
# 0
* 625
* 0
1
1
# 349
# 0
* 349
* 0
1
1
# 681
# 0
* 681
* 0
1
1
# 200
# 0
0
0
# 678
# 0
* 678
* 0
1
1
# 260
# 0
0
0
# 280
# 0
0
0
# 799
# 0
0
0
# 333
# 0
0
0
# 880
# 0
* 880
* 0
1
1
# 

# 0
* 490
* 0
1
1
# 533
# 0
* 533
* 0
1
1
# 594
# 0
0
0
# 309
# 0
0
0
# 946
# 0
* 946
* 0
1
1
# 1021
# 0
0
0
# 352
# 0
0
0
# 503
# 0
* 503
* 0
1
1
# 563
# 0
0
0
# 542
# 0
0
0
# 907
# 0
0
0
# 365
# 0
0
0
# 735
# 0
0
0
# 194
# 0
0
0
# 473
# 0
* 473
* 0
1
1
# 512
# 0
* 512
* 0
1
1
# 497
# 0
0
0
# 372
# 0
0
0
# 642
# 0
0
0
# 696
# 0
* 696
* 0
1
1
# 696
# 0
* 696
* 0
1
1
# 284
# 0
0
0
# 705
# 0
* 705
* 0
1
1
# 873
# 0
0
0
# 656
# 0
* 656
* 0
1
1
# 402
# 0
0
0
# 809
# 0
* 809
* 0
1
1
# 963
# 0
* 963
* 0
1
1
# 785
# 0
0
0
# 567
# 0
0
0
# 797
# 0
0
0
# 605
# 0
* 605
* 0
1
1
# 341
# 0
0
0
# 272
# 0
0
0
# 668
# 0
0
0
# 499
# 0
* 499
* 0
1
1
# 239
# 0
0
1
# 441
# 0
* 441
* 0
** 441
** 0
0
2
# 243
# 0
0
0
# 870
# 0
* 870
* 0
1
1
# 294
# 0
0
0
# 248
# 0
0
0
# 269
# 0
0
0
# 831
# 0
* 831
* 0
1
1
# 200
# 0
0
0
# 707
# 0
* 707
* 0
1
2
# 876
# 0
0
0
# 339
# 0
0
0
# 869
# 0
* 869
* 0
1
1
# 740
# 0
* 740
* 0
1
1
# 192
# 0
0
0
# 985
# 0
0
0
# 682
# 0
0
0
# 271
# 0
0
0
# 181
# 0
* 181
* 0
1
1
# 605
# 0
* 6

In [39]:
Exit_Reopen_Unstack_subset.loc[Exit_Reopen_Unstack_subset.ID==461388]

ID        DOB Init_IFSP_Dt    Exit_Dt  Service_Days  \
347  461388 2017-09-13   2019-05-31 2020-09-12           470   

     Actual_Service_Days       ExDT1       ExDT2 ExDT3 ExDT4 ExDT5 ExDT6  \
347                  429  09/14/2020  10/24/2019   NaN   NaN   NaN   NaN   

          ReDT1 ReDT2 ReDT3 ReDT4 ReDT5 ReDT6  
347  12/04/2019   NaN   NaN   NaN   NaN   NaN

Exit_Reopen_Unstack_subset.drop(columns=[0,192])

## NO of records updated for Actual Service days is 504

In [41]:
Exit_Reopen_Unstack_subset.loc[Exit_Reopen_Unstack_subset.Actual_Service_Days!=Exit_Reopen_Unstack_subset.Service_Days]

ID        DOB Init_IFSP_Dt    Exit_Dt  Service_Days  \
8     469243 2019-07-26   2019-10-02 2021-02-11           498   
30    469200 2019-07-13   2019-11-13 2020-09-22           314   
59    467228 2019-06-14   2019-11-14 2020-12-17           399   
62    453185 2018-11-19   2019-07-29 2020-12-28           518   
82    436627 2018-04-12   2018-07-17 2021-04-09           997   
...      ...        ...          ...        ...           ...   
2195  448232 2018-08-13   2018-12-05 2021-01-25           782   
2198  441468 2018-04-22   2018-08-22 2021-04-21           973   
2223  469136 2017-10-24   2019-10-23 2020-10-23           366   
2227  432033 2017-12-30   2018-04-25 2020-12-29           979   
2233  428532 2017-07-13   2018-01-26 2020-07-12           898   

      Actual_Service_Days       ExDT1       ExDT2 ExDT3 ExDT4 ExDT5 ExDT6  \
8                     194  04/07/2020         NaN   NaN   NaN   NaN   NaN   
30                     96  01/21/2020  09/10/2019   NaN   NaN   NaN   NaN   
59                    335  08/10/2020  09/12/2019   NaN   NaN   NaN   NaN   
62                    437  09/11/2020  01/28/2019   NaN   NaN   NaN   NaN   
82                    402  05/28/2019  11/29/2018   NaN   NaN   NaN   NaN   
...                   ...         ...         ...   ...   ...   ...   ...   
2195                  357  10/09/2020  10/18/2019   NaN   NaN   NaN   NaN   
2198                  883  11/26/2018  10/12/2018   NaN   NaN   NaN   NaN   
2223                  273  04/26/2020         NaN   NaN   NaN   NaN   NaN   
2227                  874  10/23/2018         NaN   NaN   NaN   NaN   NaN   
2233                  744  05/24/2018         NaN   NaN   NaN   NaN   NaN   

           ReDT1       ReDT2 ReDT3 ReDT4 ReDT5 ReDT6  
8     02/05/2021         NaN   NaN   NaN   NaN   NaN  
30    08/26/2020  09/23/2019   NaN   NaN   NaN   NaN  
59    10/13/2020  11/01/2019   NaN   NaN   NaN   NaN  
62    12/01/2020  07/22/2019   NaN   NaN   NaN   NaN  
82    10/06/2020  03/07/2019   NaN   NaN   NaN   NaN  
...          ...         ...   ...   ...   ...   ...  
2195  12/30/2020  09/25/2020   NaN   NaN   NaN   NaN  
2198  01/30/2019  11/06/2018   NaN   NaN   NaN   NaN  
2223  07/28/2020         NaN   NaN   NaN   NaN   NaN  
2227  02/05/2019         NaN   NaN   NaN   NaN   NaN  
2233  10/25/2018         NaN   NaN   NaN   NaN   NaN  

[504 rows x 18 columns]

In [42]:
Exit_Reopen_Subset=Exit_Reopen_Unstack_subset[['ID', 'Actual_Service_Days']]

In [43]:
df = Exit_Notes.merge(Exit_Reopen_Subset, how='left', on='ID')
print(df)


     POE      ID        DOB Init_IFSP_Dt  \
0     UC  471738 2019-02-28   2019-10-31   
1     FT  449030 2018-07-22   2018-12-18   
2     FT  463311 2017-07-09   2019-07-02   
3     FT  463313 2017-07-09   2019-07-02   
4     FT  469194 2017-08-31   2019-10-11   
...   ..     ...        ...          ...   
5577  MD  491609 2018-12-22   2020-11-09   
5578  GN  491743 2018-06-05   2020-11-09   
5579  ET  491879 2019-04-24   2020-11-04   
5580  SC  492267 2018-06-01   2020-11-24   
5581  SW  492408 2020-08-06   2020-11-20   

                                     Exit_Rsn    Exit_Dt  Service_Days  \
0                       618 - Parent withdraw 2020-09-21           326   
1                    618 - Moved out of state 2021-03-06           809   
2     618 - Part B eligibility not determined 2020-07-08           372   
3     618 - Part B eligibility not determined 2020-07-08           372   
4     618 - Part B eligibility not determined 2020-08-30           324   
...                        

In [44]:
df['Actual_Service_Days'] = df['Actual_Service_Days_y'].fillna(df['Actual_Service_Days_x'])


In [45]:
print(df)

     POE      ID        DOB Init_IFSP_Dt  \
0     UC  471738 2019-02-28   2019-10-31   
1     FT  449030 2018-07-22   2018-12-18   
2     FT  463311 2017-07-09   2019-07-02   
3     FT  463313 2017-07-09   2019-07-02   
4     FT  469194 2017-08-31   2019-10-11   
...   ..     ...        ...          ...   
5577  MD  491609 2018-12-22   2020-11-09   
5578  GN  491743 2018-06-05   2020-11-09   
5579  ET  491879 2019-04-24   2020-11-04   
5580  SC  492267 2018-06-01   2020-11-24   
5581  SW  492408 2020-08-06   2020-11-20   

                                     Exit_Rsn    Exit_Dt  Service_Days  \
0                       618 - Parent withdraw 2020-09-21           326   
1                    618 - Moved out of state 2021-03-06           809   
2     618 - Part B eligibility not determined 2020-07-08           372   
3     618 - Part B eligibility not determined 2020-07-08           372   
4     618 - Part B eligibility not determined 2020-08-30           324   
...                        

In [46]:
df.loc[df.Actual_Service_Days!=0]

POE      ID        DOB Init_IFSP_Dt  \
5     UC  473474 2017-11-05   2020-05-06   
8     GN  469243 2019-07-26   2019-10-02   
12    SE  436339 2018-04-05   2019-02-13   
16    ET  474102 2018-05-11   2020-04-09   
21    GN  476958 2018-10-17   2020-08-24   
...   ..     ...        ...          ...   
5577  MD  491609 2018-12-22   2020-11-09   
5578  GN  491743 2018-06-05   2020-11-09   
5579  ET  491879 2019-04-24   2020-11-04   
5580  SC  492267 2018-06-01   2020-11-24   
5581  SW  492408 2020-08-06   2020-11-20   

                                     Exit_Rsn    Exit_Dt  Service_Days  \
5     618 - Part B eligibility not determined 2020-11-04           182   
8                       618 - Parent withdraw 2021-02-11           498   
12    618 - Part B eligibility not determined 2021-04-01           778   
16    618 - Part B eligibility not determined 2021-05-10           396   
21                      618 - Parent withdraw 2021-05-25           274   
...                                       ...        ...           ...   
5577   618 - Attempts to contact unsuccessful 2021-05-17           189   
5578  618 - Part B eligibility not determined 2021-06-04           207   
5579                    618 - Parent withdraw 2021-05-06           183   
5580  618 - Part B eligibility not determined 2021-05-31           188   
5581   618 - Attempts to contact unsuccessful 2021-05-20           181   

      Actual_Service_Days_x  \
5                         0   
8                         0   
12                        0   
16                        0   
21                        0   
...                     ...   
5577                    189   
5578                    207   
5579                    183   
5580                    188   
5581                    181   

                                                                                                                                                                                                                                                           Exit_Note  \
5                                           **LEA county on the Transition page and demographics page do not match due to child moving out of county following the transition meeting. Reopened: 04/06/2020 Exit Date: 11/26/2019 Exit Reason: Ineligible for Part C   
8                                             02-11-2021 – PWN signed and returned for voluntary closure/declined services. Family made aware they can reopen the file if they choose. Reopened: 02/05/2021 Exit Date: 04/07/2020 Exit Reason: 618 - Parent withdraw   
12    04/01/2021- File closed business day prior to child's birthday. Parents declined TPC with LEA and plan to continue with private preschool and feeding therapy through insurance. Reopened: 01/11/2019 Exit Date: 06/17/2018 Exit Reason: Ineligible for Part C   
16                                                                                                  05/10/2021 TPC held then parents withdrew from testing and thus part B not determined. Reopened: 02/28/2020 Exit Date: 12/05/2019 Exit Reason: Unable to contact   
21                                                                                                                                                               05/25/21- parent withdraw Reopened: 08/04/2020 Exit Date: 03/09/2020 Exit Reason: Unable to contact   
...                                                                                                                                                                                                                                                              ...   
5577                                                                                                                                                                                                                                                             NaN   
5578                                                                                                  

In [47]:
df.loc[df.ID==477472]

POE      ID        DOB Init_IFSP_Dt  \
1799  ET  477472 2017-09-22   2020-03-13   

                                     Exit_Rsn    Exit_Dt  Service_Days  \
1799  618 - Part B eligibility not determined 2020-09-21           192   

      Actual_Service_Days_x  \
1799                      0   

                                                                                                                                                                                                                                                              Exit_Note  \
1799  Reopened: 09/21/2020 Exit Date: 09/21/2020 Exit Reason: 618 - Part B eligibility not determined Exit Date: 09/21/2020 Exit Reason: Child aged out of TEIS. She had TPC with LEA, and mom, reported on 9/16/2020 evals with LEA are scheduled in the coming weeks.   

      Actual_Service_Days_y  Actual_Service_Days  
1799                  192.0                192.0

In [48]:
df['Actual_Service_Days'].unique()

array([   0.,  182.,  194.,  778.,  396.,  274.,  902.,  185.,   96.,
        281.,  584.,  438.,  517.,  335.,  437.,  856.,  426.,  997.,
        402.,  620.,  164.,  371.,  184.,  932.,  661.,  372.,  785.,
        484.,  306.,  243.,  881.,  441.,  435.,  612.,  715.,  315.,
        328.,  672.,  645.,  341.,   51.,  539.,  548.,  212.,  220.,
        454.,  373.,  391.,  783.,  187.,  860.,  359.,  278.,  353.,
        238.,  554.,  307.,  271.,  651.,  658.,  518.,  653.,  348.,
        854.,  993.,  207.,  963.,  647.,  717.,  452.,  802.,  300.,
        451.,  486.,  895.,  659.,  411.,  575.,  459.,  504.,  429.,
        493.,  843.,  231.,  833.,  246.,  728.,  280.,  309.,  773.,
        490.,  245.,  292.,  261.,  277.,  217.,  520.,  256.,  668.,
        494.,  193.,  237.,  859.,  512.,  611.,  369.,  203., 1007.,
        603.,  632.,  466.,  295.,  206.,  404.,  560.,  704.,  585.,
        742.,  350.,  188.,  199.,  713.,  230.,  253.,  624.,  716.,
        145.,  940.,

## convert float into int

In [49]:
df['Actual_Service_Days'] = df['Actual_Service_Days'].astype(int)

In [50]:
df['Actual_Service_Days'].unique()

array([   0,  182,  194,  778,  396,  274,  902,  185,   96,  281,  584,
        438,  517,  335,  437,  856,  426,  997,  402,  620,  164,  371,
        184,  932,  661,  372,  785,  484,  306,  243,  881,  441,  435,
        612,  715,  315,  328,  672,  645,  341,   51,  539,  548,  212,
        220,  454,  373,  391,  783,  187,  860,  359,  278,  353,  238,
        554,  307,  271,  651,  658,  518,  653,  348,  854,  993,  207,
        963,  647,  717,  452,  802,  300,  451,  486,  895,  659,  411,
        575,  459,  504,  429,  493,  843,  231,  833,  246,  728,  280,
        309,  773,  490,  245,  292,  261,  277,  217,  520,  256,  668,
        494,  193,  237,  859,  512,  611,  369,  203, 1007,  603,  632,
        466,  295,  206,  404,  560,  704,  585,  742,  350,  188,  199,
        713,  230,  253,  624,  716,  145,  940,  797,  957,  476,  213,
        368,  847,  365,  225,  447,  264,  257,  488,  884, 1029,  233,
        992,  528,  765,  894,  374,  407,  279,  7

In [51]:
df=df.drop(['Actual_Service_Days_x', 'Actual_Service_Days_y'], axis=1)

## For Analysis

In [52]:
df_analysis=df.loc[df.Actual_Service_Days==0]

df_analysis.to_csv("../data/df_analysis.csv", index = False)

In [54]:
print(df_analysis)

     POE      ID        DOB Init_IFSP_Dt  \
0     UC  471738 2019-02-28   2019-10-31   
1     FT  449030 2018-07-22   2018-12-18   
2     FT  463311 2017-07-09   2019-07-02   
3     FT  463313 2017-07-09   2019-07-02   
4     FT  469194 2017-08-31   2019-10-11   
...   ..     ...        ...          ...   
2246  GN  463962 2017-07-02   2019-07-10   
2247  GN  454305 2017-07-06   2019-03-08   
2248  GN  450369 2017-08-16   2018-12-19   
2249  GN  476783 2017-11-24   2020-03-04   
2250  MD  486852 2020-04-10   2020-09-09   

                                                   Exit_Rsn    Exit_Dt  \
0                                     618 - Parent withdraw 2020-09-21   
1                                  618 - Moved out of state 2021-03-06   
2                   618 - Part B eligibility not determined 2020-07-08   
3                   618 - Part B eligibility not determined 2020-07-08   
4                   618 - Part B eligibility not determined 2020-08-30   
...                        

In [55]:
df_analysis['Actual_Service_Days'] = df_analysis.apply(lambda x: x['Service_Days'] if x['Actual_Service_Days'] == 0 else x['Actual_Service_Days'], axis=1)

C:\Users\Anand\AppData\Local\Temp/ipykernel_21912/4053232548.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_analysis['Actual_Service_Days'] = df_analysis.apply(lambda x: x['Service_Days'] if x['Actual_Service_Days'] == 0 else x['Actual_Service_Days'], axis=1)


In [56]:
df.loc[df.Actual_Service_Days==0]

POE      ID        DOB Init_IFSP_Dt  \
0     UC  471738 2019-02-28   2019-10-31   
1     FT  449030 2018-07-22   2018-12-18   
2     FT  463311 2017-07-09   2019-07-02   
3     FT  463313 2017-07-09   2019-07-02   
4     FT  469194 2017-08-31   2019-10-11   
...   ..     ...        ...          ...   
2246  GN  463962 2017-07-02   2019-07-10   
2247  GN  454305 2017-07-06   2019-03-08   
2248  GN  450369 2017-08-16   2018-12-19   
2249  GN  476783 2017-11-24   2020-03-04   
2250  MD  486852 2020-04-10   2020-09-09   

                                                   Exit_Rsn    Exit_Dt  \
0                                     618 - Parent withdraw 2020-09-21   
1                                  618 - Moved out of state 2021-03-06   
2                   618 - Part B eligibility not determined 2020-07-08   
3                   618 - Part B eligibility not determined 2020-07-08   
4                   618 - Part B eligibility not determined 2020-08-30   
...                                                     ...        ...   
2246                618 - Part B eligibility not determined 2020-07-01   
2247  618 - Completion of IFSP/ no longer requires services 2020-07-02   
2248                618 - Part B eligibility not determined 2020-08-15   
2249                618 - Part B eligibility not determined 2020-11-23   
2250                 618 - Attempts to contact unsuccessful 2021-04-16   

      Service_Days  \
0              326   
1              809   
2              372   
3              372   
4              324   
...            ...   
2246           357   
2247           482   
2248           605   
2249           264   
2250           219   

                                                                                                                                                                                                                                                        Exit_Note  \
0                                                                                                                                                                                                                         * No Exit ECO available due to COVID-19   
1                                                                                     ***family moving out of state March 2021, and will complete the Part B eligibility testing in their new state; SC provided EI program information for Missouri to family***   
2     **Completed the transition planning conference via 3 way call with family, SC, and LEA due to school closure and the impact of the COVID-19 pandemic. An assessment is TBD. SPED Prek teacher will follow up with family to schedule once school resumes.**   
3     **Completed the transition planning conference via 3 way call with family, SC, and LEA due to school closure and the impact of the COVID-19 pandemic. An assessment is TBD. SPED Prek teacher will follow up with family to schedule once school resumes.**   
4                                                                                                       **Due to the impact of the COVID-19 pandemic and school closure an assessment will be scheduled once school resumes. LEA will contact family to schedule.   
...                                                                                                                                                                                                                                                           ...   
2246                                                                                                                                                                                        Transition meeting late due to COVID-19 Pandemic and School closings.   
2247                                                                                                                                           Transition meeting late due to COVID-19 Pandemic and School closings. No new eco scores due

## Actual Service days= service days for children with  exit notes having none of exit or reopened dates

In [57]:
df['Actual_Service_Days'] = df.apply(lambda x: x['Service_Days'] if x['Actual_Service_Days'] == 0 else x['Actual_Service_Days'], axis=1)

In [58]:
Exit_Notes.loc[Exit_Notes.Service_Days<182]

POE      ID        DOB Init_IFSP_Dt  \
383   MD  480170 2017-12-22   2020-06-24   
426   MD  486942 2018-10-21   2020-09-11   
432   GN  474794 2017-07-10   2020-01-10   
502   FT  476751 2019-12-16   2020-03-04   
1018  NW  477159 2018-12-17   2020-09-09   
1120  SC  480223 2018-09-12   2020-05-14   
1172  MD  458186 2017-11-27   2020-05-29   
1374  ET  440527 2018-06-20   2020-07-14   
1692  UC  482147 2020-01-09   2020-09-30   
1936  GN  480514 2020-03-11   2020-06-19   
1971  ET  487833 2019-09-19   2020-12-03   
1972  ET  487832 2019-09-19   2020-12-03   
2129  SC  465241 2017-07-20   2020-01-21   
2142  SC  468796 2019-07-27   2020-01-24   
2186  SE  482693 2018-12-07   2020-06-18   
4579  MD  473838 2018-11-16   2020-01-30   
4706  MD  474730 2019-02-25   2020-02-06   
4849  UC  476252 2019-04-08   2020-01-28   
4910  ET  476841 2018-01-18   2020-02-21   
5151  ET  479854 2019-10-05   2020-06-03   
5202  SC  481129 2018-12-27   2020-06-04   
5324  ET  482995 2017-12-27   2020-06-29   
5332  SW  483137 2020-03-19   2020-07-09   
5393  GN  484003 2018-09-26   2020-07-17   
5441  FT  485198 2018-10-03   2020-07-31   
5468  GN  485953 2018-02-03   2020-08-06   
5472  ET  486048 2019-10-08   2020-08-27   
5537  SC  488453 2020-06-27   2020-10-07   
5571  FT  490726 2018-05-04   2020-11-04   
5581  SW  492408 2020-08-06   2020-11-20   

                                                   Exit_Rsn    Exit_Dt  \
383                 618 - Part B eligibility not determined 2020-12-21   
426                                   618 - Parent withdraw 2021-03-11   
432                 618 - Part B eligibility not determined 2020-07-09   
502                  618 - Attempts to contact unsuccessful 2020-08-31   
1018                               618 - Moved out of state 2021-03-08   
1120                                  618 - Parent withdraw 2020-11-10   
1172                                  618 - Part B eligible 2020-11-26   
1374                                  618 - Parent withdraw 2021-01-11   
1692                                  618 - Parent withdraw 2021-03-30   
1936                 618 - Attempts to contact unsuccessful 2020-12-17   
1971                                  618 - Parent withdraw 2021-06-01   
1972                                  618 - Parent withdraw 2021-06-01   
2129                618 - Part B eligibility not determined 2020-07-19   
2142                                  618 - Parent withdraw 2020-07-22   
2186                                  618 - Parent withdraw 2020-12-16   
4579  618 - Completion of IFSP/ no longer requires services 2020-07-29   
4706                                  618 - Parent withdraw 2020-08-04   
4849                 618 - Attempts to contact unsuccessful 2020-07-27   
4910                                  618 - Parent withdraw 2020-08-20   
5151                                  618 - Parent withdraw 2020-12-01   
5202                                  618 - Parent withdraw 2020-12-02   
5324                                  618 - Part B eligible 2020-12-26   
5332                                  618 - Parent withdraw 2021-01-05   
5393                 618 - Attempts to contact unsuccessful 2021-01-13   
5441                 618 - Attempts to contact unsuccessful 2021-01-28   
5468                                  618 - Part B eligible 2021-02-02   
5472                                  618 - Parent withdraw 2021-02-24   
5537                 618 - Attempts to contact unsuccessful 2021-04-05   
5571                                  618 - Part B eligible 2021-05-03   
5581                 618 - Attempts to contact unsuccessful 2021-05-20   

      Service_Days  Actual_Service_Days  \
383            180                    0   
426            181                    0   
432            181                    0   
502            180                    0   
1018           180                    0   
1120           180                    0   
1172           181                    0   
1

In [59]:
df.loc[df.Actual_Service_Days<182]

POE      ID        DOB Init_IFSP_Dt  \
30    GN  469200 2019-07-13   2019-11-13   
85    NW  475485 2019-09-02   2020-10-22   
178   MD  439318 2018-06-24   2020-02-18   
383   MD  480170 2017-12-22   2020-06-24   
426   MD  486942 2018-10-21   2020-09-11   
...   ..     ...        ...          ...   
5468  GN  485953 2018-02-03   2020-08-06   
5472  ET  486048 2019-10-08   2020-08-27   
5537  SC  488453 2020-06-27   2020-10-07   
5571  FT  490726 2018-05-04   2020-11-04   
5581  SW  492408 2020-08-06   2020-11-20   

                                     Exit_Rsn    Exit_Dt  Service_Days  \
30     618 - Attempts to contact unsuccessful 2020-09-22           314   
85                      618 - Parent withdraw 2021-05-25           215   
178    618 - Attempts to contact unsuccessful 2020-09-01           196   
383   618 - Part B eligibility not determined 2020-12-21           180   
426                     618 - Parent withdraw 2021-03-11           181   
...                                       ...        ...           ...   
5468                    618 - Part B eligible 2021-02-02           180   
5472                    618 - Parent withdraw 2021-02-24           181   
5537   618 - Attempts to contact unsuccessful 2021-04-05           180   
5571                    618 - Part B eligible 2021-05-03           180   
5581   618 - Attempts to contact unsuccessful 2021-05-20           181   

                                                                                                                                                                                                                                                                                                                                                                                                                                  Exit_Note  \
30                                                                                                                                                                                                                                                    09/22/2020- unable to contact Reopened: 08/26/2020 Exit Date: 01/21/2020 Exit Reason: 618 - Parent withdraw Reopened: 09/23/2019 Exit Date: 09/10/2019 Exit Reason: Unable to contact   
85                                                                                                                                                                            5/25/2021- Family declined reopening file. See CL for notes. Reopened: 05/20/2021 Exit Date: 03/30/2021 Exit Reason: 618 - Parent withdraw 3/30/2021- Closed file. See CL for notes Reopened: 10/06/2020 Exit Date: 01/09/2020 Exit Reason: Unable to contact   
178   Close date: 09/01/2020 Close Reason: attempts to contact unsuccessful Reopened: 07/13/2020 Exit Date: 02/19/2020 Exit Reason: 618 - Parent withdraw Close Date: 02/19/2020 Close Reason: Parent withdrawal Reopened: 01/07/2020 Exit Date: 02/01/2019 Exit Reason: Unable to contact Reopened: 01/17/2019 Exit Date: 01/08/2019 Exit Reason: Unable to contact Reopened: 12/12/2018 Exit Date: 07/06/2018 Exit Reason: Parent decline   
383                                                                                                                                                                                                                                                                                                                                                                    Exit Date: 12/21/2020 Exit Reason: Part B Eligibility not Determined   
426                                                                                                                                                                                                                                                                                                                                                                          Exit/Close Date: 03/11/2021 Exit/Close Reason: Parent Withdraw   
.

In [83]:
Exit_Notes_Actual=Exit_Notes_Original.merge(df[['ID','Actual_Service_Days']],how='inner',left_on='Child ID',right_on='ID')

In [85]:
 Exit_Notes_Actual=Exit_Notes_Actual.drop(columns='ID')


In [92]:
Exit_Notes_Actual.loc[Exit_Notes_Actual.Actual_Service_Days<182].shape

(102, 12)

In [96]:
Exit_Notes_Original.loc[Exit_Notes_Original['Days from I-IFSP to Exit'] <182].shape

(30, 11)

In [97]:
Exit_Notes_Actual.loc[Exit_Notes_Actual.Actual_Service_Days != Exit_Notes_Original['Days from I-IFSP to Exit'] ].shape

(504, 12)

In [104]:
ECO_with_Exit20_21_Actual=ECO_with_Exit20_21_Original.merge(Exit_Notes_Actual[['Child ID','Actual_Service_Days']],how='left',left_on='CHILD_ID',right_on='Child ID')
#ECO_with_Exit20_21_Original.head()

In [105]:
ECO_with_Exit20_21_Actual.head()

DISTRICT CHILD_ID ECO_Entry_DATE  Ent SOCIAL_SCALE  Ent KNOWLEDGE_SCALE  \
0       ET   456849     2019-04-24               3.0                  5.0   
1       ET   490916            NaT               NaN                  NaN   
2       ET   426757     2019-03-13               2.0                  3.0   
3       ET   480611     2020-05-28               2.0                  5.0   
4       ET   492863            NaT               NaN                  NaN   

   Ent APPROPRIATE_ACTION_SCALE ECO_Exit_DATE  Exit SOCIAL_SCALE  \
0                           7.0    2019-10-14                7.0   
1                           NaN           NaT                NaN   
2                           4.0    2020-08-12                6.0   
3                           1.0    2020-11-09                5.0   
4                           NaN           NaT                NaN   

   Exit KNOWLEDGE_SCALE  Exit APPROPRIATE_ACTION_SCALE  ... OC2 Data Check  \
0                   5.0                            7.0  ...            1.0   
1                   NaN                            NaN  ...            0.0   
2                   7.0                            7.0  ...            1.0   
3                   6.0                            3.0  ...            1.0   
4                   NaN                            NaN  ...            0.0   

  OC3 - a OC3 - b OC3 - c OC3 - d OC3 - e OC3 Data Check  \
0     NaN     NaN     NaN     NaN     1.0            1.0   
1     NaN     NaN     NaN     NaN     NaN            0.0   
2     NaN     NaN     NaN     1.0     NaN            1.0   
3     NaN     NaN     1.0     NaN     NaN            1.0   
4     NaN     NaN     NaN     NaN     NaN            0.0   

  Days btw Initial and Exit  Child ID  Actual_Service_Days  
0                     569.0  456849.0                569.0  
1                     221.0       NaN                  NaN  
2                     533.0  426757.0                533.0  
3                     298.0  480611.0                298.0  
4                     147.0       NaN                  NaN  

[5 rows x 71 columns]

In [107]:
ECO_with_Exit20_21_Actual=ECO_with_Exit20_21_Actual.drop(columns='Child ID')

In [128]:
ECO_with_Exit20_21_Actual.tail(10)

DISTRICT CHILD_ID ECO_Entry_DATE  Ent SOCIAL_SCALE  Ent KNOWLEDGE_SCALE  \
8431       UC   446579     2018-10-27               1.0                  1.0   
8432       UC   431646     2018-07-10               2.0                  4.0   
8433       UC   430864     2018-03-23               1.0                  1.0   
8434       UC   430862     2018-03-23               1.0                  1.0   
8435       UC   482155            NaT               NaN                  NaN   
8436       FT   422823            NaT               NaN                  NaN   
8437       UC   457024     2019-04-05               5.0                  5.0   
8438       UC   496398            NaT               NaN                  NaN   
8439       UC   486316            NaT               NaN                  NaN   
8440       UC   472195            NaT               NaN                  NaN   

      Ent APPROPRIATE_ACTION_SCALE ECO_Exit_DATE  Exit SOCIAL_SCALE  \
8431                           2.0    2019-10-11                6.0   
8432                           1.0    2020-02-26                7.0   
8433                           1.0    2020-03-09                7.0   
8434                           1.0    2020-03-09                7.0   
8435                           NaN           NaT                NaN   
8436                           NaN           NaT                NaN   
8437                           5.0    2020-03-12                6.0   
8438                           NaN           NaT                NaN   
8439                           NaN           NaT                NaN   
8440                           NaN           NaT                NaN   

      Exit KNOWLEDGE_SCALE  Exit APPROPRIATE_ACTION_SCALE  ... OC2 - e  \
8431                   5.0                            7.0  ...     NaN   
8432                   4.0                            5.0  ...     NaN   
8433                   7.0                            7.0  ...     NaN   
8434                   7.0                            7.0  ...     NaN   
8435                   NaN                            NaN  ...     NaN   
8436                   NaN                            NaN  ...     NaN   
8437                   3.0                            4.0  ...     NaN   
8438                   NaN                            NaN  ...     NaN   
8439                   NaN                            NaN  ...     NaN   
8440                   NaN                            NaN  ...     NaN   

     OC2 Data Check OC3 - a OC3 - b OC3 - c OC3 - d OC3 - e OC3 Data Check  \
8431            1.0     NaN     NaN     NaN     1.0     NaN            1.0   
8432            1.0     NaN     NaN     1.0     NaN     NaN            1.0   
8433            1.0     NaN     NaN     NaN     1.0     NaN            1.0   
8434            1.0     NaN     NaN     NaN     1.0     NaN            1.0   
8435            0.0     NaN     NaN     NaN     NaN     NaN            0.0   
8436            0.0     NaN     NaN     NaN     NaN     NaN            0.0   
8437            1.0     NaN     1.0     NaN     NaN     NaN            1.0   
8438            0.0     NaN     NaN     NaN     NaN     NaN            0.0   
8439            0.0     NaN     NaN     NaN     NaN     NaN            0.0   
8440            0.0     NaN     NaN     NaN     NaN     NaN            0.0   

      Days btw Initial and Exit  Actual_Service_Days  
8431                      731.0                731.0  
8432                      818.0                818.0  
8433                      914.0                914.0  
8434                      914.0                914.0  
8435                      112.0                  NaN  
8436                     1012.0                 12.0  
8437                      484.0                484.0  
8438                       91.0                  NaN  
8439                      218.0                218.0  
8440                      422.0                422.0  

[10 rows x 70 columns]

In [127]:
ECO_with_Exit20_21_Actual.drop(ECO_with_Exit20_21_Actual.tail(5).index,
        inplace = True)

In [132]:
ECO_with_Exit20_21_Actual['Days btw Initial and Exit'] = ECO_with_Exit20_21_Actual['Days btw Initial and Exit'].astype(int)
ECO_with_Exit20_21_Actual['Actual_Service_Days']=ECO_with_Exit20_21_Actual['Actual_Service_Days'].fillna(ECO_with_Exit20_21_Actual['Days btw Initial and Exit'])
ECO_with_Exit20_21_Actual['Actual_Service_Days'] = ECO_with_Exit20_21_Actual['Actual_Service_Days'].astype(int)

ECO_with_Exit20_21.to_csv('../data/Eco_Analysis.csv',index=False)

In [133]:
ECO_with_Exit20_21_Actual.loc[ECO_with_Exit20_21_Actual['Actual_Service_Days'] ==0]

DISTRICT CHILD_ID ECO_Entry_DATE  Ent SOCIAL_SCALE  Ent KNOWLEDGE_SCALE  \
11        GN   486884            NaT               NaN                  NaN   
34        MD   492159            NaT               NaN                  NaN   
38        NW   487712            NaT               NaN                  NaN   
56        NW   490893            NaT               NaN                  NaN   
62        SC   431258            NaT               NaN                  NaN   
64        SC   489306            NaT               NaN                  NaN   
71        SW   495449            NaT               NaN                  NaN   
72        SW   498252            NaT               NaN                  NaN   
106       SW   487030            NaT               NaN                  NaN   
110       SW   502756            NaT               NaN                  NaN   

     Ent APPROPRIATE_ACTION_SCALE ECO_Exit_DATE  Exit SOCIAL_SCALE  \
11                            NaN           NaT                NaN   
34                            NaN           NaT                NaN   
38                            NaN           NaT                NaN   
56                            NaN           NaT                NaN   
62                            NaN           NaT                NaN   
64                            NaN           NaT                NaN   
71                            NaN           NaT                NaN   
72                            NaN           NaT                NaN   
106                           NaN           NaT                NaN   
110                           NaN           NaT                NaN   

     Exit KNOWLEDGE_SCALE  Exit APPROPRIATE_ACTION_SCALE  ... OC2 - e  \
11                    NaN                            NaN  ...     NaN   
34                    NaN                            NaN  ...     NaN   
38                    NaN                            NaN  ...     NaN   
56                    NaN                            NaN  ...     NaN   
62                    NaN                            NaN  ...     NaN   
64                    NaN                            NaN  ...     NaN   
71                    NaN                            NaN  ...     NaN   
72                    NaN                            NaN  ...     NaN   
106                   NaN                            NaN  ...     NaN   
110                   NaN                            NaN  ...     NaN   

    OC2 Data Check OC3 - a OC3 - b OC3 - c OC3 - d OC3 - e OC3 Data Check  \
11             0.0     NaN     NaN     NaN     NaN     NaN            0.0   
34             0.0     NaN     NaN     NaN     NaN     NaN            0.0   
38             0.0     NaN     NaN     NaN     NaN     NaN            0.0   
56             0.0     NaN     NaN     NaN     NaN     NaN            0.0   
62             0.0     NaN     NaN     NaN     NaN     NaN            0.0   
64             0.0     NaN     NaN     NaN     NaN     NaN            0.0   
71             0.0     NaN     NaN     NaN     NaN     NaN            0.0   
72             0.0     NaN     NaN     NaN     NaN     NaN            0.0   
106            0.0     NaN     NaN     NaN     NaN     NaN            0.0   
110            0.0     NaN     NaN     NaN     NaN     NaN            0.0   

     Days btw Initial and Exit  Actual_Service_Days  
11                           0                    0  
34                           0                    0  
38                           0                    0  
56                           0                    0  
62                           0                    0  
64                           0                    0  
71                           0                    0  
72                           0                    0  
106                          0                    0  
110                          0                    0  

[10 rows x 70 columns]